bot/exts/moderation/voice_gate.py

```diff
@@ -160,6 +186,35 @@ async def on_message(self, message: discord.Message) -> None:
         with suppress(discord.NotFound):
             await message.delete()
 
+    @Cog.listener()
+    async def on_voice_state_update(self, member: Member, *_) -> None:
+        """Pings a user if they've never joined the voice chat before and aren't verified."""
+        if member.bot:
+            log.trace("User is a bot. Ignore.")
+            return
+
+        in_cache = await self.redis_cache.get(member.id, None)
+
+        # member.voice will return None if the user is not in a voice channel
+        if in_cache is None and member.voice is not None:
+            log.trace("User not in cache and is in a voice channel")
+            verified = any(Roles.voice_verified == role.id for role in member.roles)
+            if verified:
+                log.trace("User is verified, add to the cache and ignore")
+                #  redis cache does not accept None, so False is used to signify no message
+                await self.redis_cache.set(member.id, False)
+                return
+
+            log.trace("User is unverified. Send ping.")
+            message = await self._voice_verification_channel.send(VOICE_PING.format(member.mention))
+            await self.redis_cache.set(member.id, message.id)
+
+            # Message will try to be deleted after 1 minutes. If it fails, it'll do so silently
+            await message.delete(delay=GateConf.voice_ping_delete_delay)
+        else:
+            log.trace("User is either in the cache or not in a voice channel. Ignore.")
+            return
```

Comment:
The return at the end of the function feels awkward. Could flip the conditionals and return early like we do with the `member.bot` check, I'd probably find that cleaner (it's the same pattern of `check` - `log` - `return`).

Reply:
Restructured with this in mind in latest commit.